src/hotspot/share/ci/ciArrayKlass.cpp

```diff
@@ -104,10 +105,36 @@ bool ciArrayKlass::is_leaf_type() {
 ciArrayKlass* ciArrayKlass::make(ciType* element_type) {
   if (element_type->is_primitive_type()) {
     return ciTypeArrayKlass::make(element_type->basic_type());
-  } else if (element_type->flatten_array()) {
-    return ciFlatArrayKlass::make(element_type->as_klass());
   } else {
-    return ciObjArrayKlass::make(element_type->as_klass());
+    return make(element_type->as_klass(), element_type->is_null_free());
+  }
+}
+
+ciArrayKlass* ciArrayKlass::make(ciKlass* klass, bool null_free) {
+  if (null_free) {
+    if (klass->is_loaded()) {
+      bool is_array_flattened = false;
+      GUARDED_VM_ENTRY(
+        EXCEPTION_CONTEXT;
+        Klass* ak = InlineKlass::cast(klass->get_Klass())->null_free_inline_array_klass(THREAD);
+        if (HAS_PENDING_EXCEPTION) {
+          CLEAR_PENDING_EXCEPTION;
+        } else {
+          if (ak != NULL && ak->is_flatArray_klass()) {
+            is_array_flattened = true;
```

Comment:
Why is that boolean needed? Can't we simply return `ciFlatArrayKlass::make(klass)` here?

Reply:
Boolean removed in the latest commit.